File: Discord_win/Discord_win/Tools/ResourcesCreator.cs
using Discord.Models;
using Discord.Resources.Static;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;

namespace Discord.Tools {
    public static class ResourcesCreator {
        //private static APICaller apiCaller;
        public static void Establish() {
            //apiCaller = new APICaller();
        }
        public static async Task<ICollection<Server>> GetListServerAsync(int userId) {
            string requestUrl = Route.Server.BuildGetByUserUrl(Inventory.CurrentUser.UserId);
            HttpResponseMessage httpResponseMessage = await new APICaller(HttpMethod.Get, requestUrl).SendRequestAsync();
            string result = await httpResponseMessage.Content.ReadAsStringAsync();
            return JsonConvert.DeserializeObject<ICollection<Server>>(result);
        }
        public static async Task<ICollection<Channel>> GetListChannelAsync(int serverId) {
            string requestUrl = Route.Channel.BuildGetByServerUrl(serverId);
            HttpResponseMessage httpResponseMessage = await new APICaller(HttpMethod.Get, requestUrl).SendRequestAsync();
            string result = await httpResponseMessage.Content.ReadAsStringAsync();
            return JsonConvert.DeserializeObject<ICollection<Channel>>(result);
        }
        public static async Task<ICollection<User>> GetListUserAsync(int serverId) {
            string requestUrl = Route.User.BuildGetByServer(serverId);
            HttpResponseMessage httpResponseMessage = await new APICaller(HttpMethod.Get, requestUrl).SendRequestAsync();
            string result = await httpResponseMessage.Content.ReadAsStringAsync();
            return JsonConvert.DeserializeObject<ICollection<User>>(result);
        }
        public static async Task<IList<Role>> GetListRoleAsync(int serverId) {
            string requestUrl = Route.Role.BuildGetByServerUrl(serverId);
            HttpResponseMessage httpResponseMessage = await new APICaller(HttpMethod.Get, requestUrl).SendRequestAsync();
            string result = await httpResponseMessage.Content.ReadAsStringAsync();
            return JsonConvert.DeserializeObject<IList<Role>>(result);
        }
        public static async Task<ICollection<Message>> GetListMessageAsync(int channelId) {
            string requestUrl = Route.Message.BuildGetByChannelUrl(channelId);
            HttpResponseMessage httpResponseMessage = await new APICaller(HttpMethod.Get, requestUrl).SendRequestAsync();
            string result = await httpResponseMessage.Content.ReadAsStringAsync();
            return JsonConvert.DeserializeObject<ICollection<Message>>(result);
        }
        public static async Task<Server> GetServerByInstantInviteAsync(string instantInvite) {
            string requestUrl = Route.InstantInvite.BuildGetServerUrl(Inventory.CurrentUser.UserId, instantInvite);
            HttpResponseMessage httpResponseMessage = await new APICaller(HttpMethod.Get, requestUrl).SendRequestAsync();
            if (!httpResponseMessage.IsSuccessStatusCode) {
                return null;
            }
            string result = await httpResponseMessage.Content.ReadAsStringAsync();
            return JsonConvert.DeserializeObject<Server>(result);
        }
        public static async Task<string> GetInstantInviteByServerAsync(int serverId) {
            string requestUrl = Route.InstantInvite.BuildGetByServerUrl(serverId);
            HttpResponseMessage httpResponseMessage = await new APICaller(HttpMethod.Get, requestUrl).SendRequestAsync();
            if (!httpResponseMessage.IsSuccessStatusCode) {
                return null;
            }
            return await httpResponseMessage.Content.ReadAsStringAsync();
        }
        public static async Task<Server> CreateServerAsync(string serverName) {
            string requestUrl = Route.Server.UrlAdd;
            APICaller apiCaller = new APICaller();
            apiCaller.SetProperties(HttpMethod.Post, requestUrl, new Server { AdminId = Inventory.CurrentUser.UserId, ServerName = serverName });
            HttpResponseMessage httpResponseMessage = await apiCaller.SendRequestAsync();
            string result = await httpResponseMessage.Content.ReadAsStringAsync();
            return JsonConvert.DeserializeObject<Server>(result);
        }
        public static async Task<Role> GetUserRoleInCurrentServerAsync(int userId, int serverId) {
            string requestUrl = Route.Role.BuildRouteGetUserRoleInServerUrl(userId, serverId);
            HttpResponseMessage httpResponseMessage = await new APICaller(HttpMethod.Get, requestUrl).SendRequestAsync();
            string result = await httpResponseMessage.Content.ReadAsStringAsync();
            return JsonConvert.DeserializeObject<Role>(result);
        }
        public static async Task<ChannelPermission> GetChannelPermissionAsync(int channelId, int roleId) {
            string requestUrl = Route.ChannelPermission.BuildGetUrl(channelId, roleId);
            HttpResponseMessage httpResponseMessage = await new APICaller(HttpMethod.Get, requestUrl).SendRequestAsync();
            string result = await httpResponseMessage.Content.ReadAsStringAsync();
            return JsonConvert.DeserializeObject<ChannelPermission>(result);
        }
    }
}

File: Discord_android/Discord/app/src/main/java/com/peanut/discord/equipment/ResourcesCreator.java
package com.peanut.discord.equipment;
import com.google.gson.reflect.TypeToken;
import com.peanut.discord.MainActivity;
import com.peanut.discord.models.Channel;
import com.peanut.discord.models.Message;
import com.peanut.discord.models.Server;
import com.peanut.discord.models.User;
import com.peanut.discord.resources.Route;
import com.peanut.discord.tools.APICaller;

import java.lang.reflect.Type;
import java.util.List;
public final class ResourcesCreator {
    private ResourcesCreator(){}
    public static void getListServer(int userId, ActionResult<List<Server>> actionResult){
        String requestUrl = Route.Server.buildGetByUserUrl(userId);
        APICaller apiCaller = new APICaller(APICaller.RequestMethod.GET, requestUrl);
        apiCaller.setOnSuccessListener((connection, response) -> {
            Type type = new TypeToken<List<Server>>(){}.getType();
            List<Server> servers = MainActivity.gson.fromJson(response, type);
            actionResult.action(servers);
        });
        apiCaller.sendRequest();
    }
    public static void getListUser(int serverId, ActionResult<List<User>> actionResult){
        String requestUrl = Route.User.buildGetByServerUrl(serverId);
        APICaller apiCaller = new APICaller(APICaller.RequestMethod.GET, requestUrl);
        apiCaller.setOnSuccessListener((connection, response) -> {
            Type type = new TypeToken<List<User>>(){}.getType();
            List<User> users = MainActivity.gson.fromJson(response, type);
            actionResult.action(users);
        });
        apiCaller.sendRequest();
    }
    public static void getListChannel(int serverId, ActionResult<List<Channel>> actionResult){
        String requestUrl = Route.Channel.buildGetByServerUrl(serverId);
        APICaller apiCaller = new APICaller(APICaller.RequestMethod.GET, requestUrl);
        apiCaller.setOnSuccessListener((connection, response) -> {
            Type type = new TypeToken<List<Channel>>(){}.getType();
            List<Channel> channels = MainActivity.gson.fromJson(response, type);
            actionResult.action(channels);
        });
        apiCaller.sendRequest();
    }
    public static void getListMessage(int channelId, ActionResult<List<Message>> actionResult){
        String requestUrl = Route.Message.buildGetByChannelUrl(channelId);
        APICaller apiCaller = new APICaller(APICaller.RequestMethod.GET, requestUrl);
        apiCaller.setOnSuccessListener((connection, response) -> {
            Type type = new TypeToken<List<Message>>(){}.getType();
            List<Message> messages = MainActivity.gson.fromJson(response, type);
            actionResult.action(messages);
        });
        apiCaller.sendRequest();
    }
    public interface ActionResult<T>{
        void action(T result);
    }
}
